trl/trainer/accelerate_ppo.py

```diff
@@ -0,0 +1,542 @@
+            else:
+                response = self.model.generate(query_tensors[i].unsqueeze(dim=0),
+                                            max_new_tokens=gen_len, **gen_kwargs)
+            response_tensors.append(response.squeeze()[-gen_len:])
+        return response_tensors
+
+    def _build_models_and_tokenizer(
+        self, 
+        model: Optional[nn.Module] = None, 
+        ref_model: Optional[nn.Module] = None , 
+        tokenizer: Optional[PreTrainedTokenizer] = None
+    ):
+        """
+        Build models for training. This builds the reference model
+        together with the model to be trained.
+
+        Args:
+            model (`nn.Module`, *optional*):
+                The model to be trained. If `None` is passed, the model will be loaded from
+                the pretrained model specified in the config.
+            ref_model (`nn.Module`, *optional*): 
+                The reference model. If `None` is passed, the model will be loaded from
+                the pretrained model specified in the config. For mixed precision setup, 
+                the reference model will be loaded in half precision according to the 
+                `accelerator.mixed_precision` flag to save memory.
+            tokenizer (`transformers.PreTrainedTokenizer`, *optional*): 
+                The tokenizer to be used. If `None` is passed, the tokenizer will be loaded from
+                the pretrained model specified in the config.
+        """
+        target_dtype_dict = {
+            "fp16": torch.float16,
+            "bf16": torch.bfloat16,
+        }
+        target_dtype = torch.float32 if self.accelerator.mixed_precision not in target_dtype_dict else target_dtype_dict[self.accelerator.mixed_precision]
+
+        # We don't cast the base model that is trained since autocast will do it under the hood for us
+        # Check: https://discuss.pytorch.org/t/valueerror-attemting-to-unscale-fp16-gradients/81372/4
+        if model is None:
+            self.model = AutoModelForCausalLMWithValueHead.from_pretrained(self.config['model_name'])
+        else:
+            self.model = model
+        # But we can cast the reference model since the weights are not updated - we do that to save memory
+        if ref_model is None:
+            self.ref_model = AutoModelForCausalLMWithValueHead.from_pretrained(self.config['model_name'], torch_dtype=target_dtype)
+        else:
+            self.ref_model = ref_model
+
+        # tokenizer
+        if tokenizer is None:
+            self.tokenizer = AutoTokenizer.from_pretrained(self.config['model_name'])
+            # HACK: do we really need this?
+            self.tokenizer.pad_token = self.tokenizer.eos_token
+        else:
+            self.tokenizer = tokenizer
+
+
+    def step(self, 
+        queries: List[torch.LongTensor], 
+        responses: List[torch.LongTensor], 
+        scores: List[torch.FloatTensor]):
```

Comment:
I think we need to wrap those in a dataloader and run them through `prepare`, otherwise we the same query/responses/scores on all machines, no?